### PR TITLE
Set the volume mount's readonly annotation based on the ISVC annotation

### PR DESCRIPTION
**Pull request by hdefazio**
**What this PR does / why we need it**:
Fixes #3687
https://issues.redhat.com/browse/RHOAIENG-8288

**Type of changes**
 Bug fix (non-breaking change which fixes an issue)

**Feature/Issue validation/testing**:

Please describe the tests that you ran to verify your changes and relevant result summary. Provide instructions so it can be reproduced.
Please also list any relevant details for your test configuration.

Test Setup:
- Created a cluster
- Installed Service Mesh, Serverless, Opendatahub operators (in that order)
- Created a DSCI and waited for it to be Ready
- Created a DSC and waited for it to be Ready
- Scaled deployments -> opendatahub-operator-controller-manager down to zero in the openshift-operators namespace
- Replaced the image in deployments -> kserve-controller  with my local image in the  opendatahub namespace
- Created a serving runtime via yaml config
- Created an ISVC via a yaml config

Test Cases:
storage.kserve.io/readonly: "false"
readonly_false

storage.kserve.io/readonly: "true"
readonly_true

storage.kserve.io/readonly was not provided
readonly_unset

**Special notes for your reviewer**:

1. Please confirm that if this PR changes any image versions, then that's the sole change this PR makes.

**Checklist**:

- [X] Have you added unit/e2e tests that prove your fix is effective or that this feature works?
- [x] Has code been commented, particularly in hard-to-understand areas?
- [ ] Have you made corresponding changes to the documentation?

**Release note**:

```release-note

```

**Re-running failed tests**

- `/rerun-all` - rerun all failed workflows.
- `/rerun-workflow <workflow name>` - rerun a specific failed workflow. Only one workflow name can be specified. Multiple /rerun-workflow commands are allowed per comment.

### CODE REVIEW

File: pkg/webhook/admission/pod/storage_initializer_injector.go [hunk 1]
```diff
@@ -243,6 +243,22 @@ func (mi *StorageInitializerInjector) InjectStorageInitializer(pod *v1.Pod) erro
...
+
+				if err != nil {
+					return err
+				}
+			}
+
+			// If StorageReadonly is set to false, the readOnly field in the volume mount's annotation is also set to false
+			// Otherwise, if it is true or unset, use the default behavior
+			readonly := true
+			if isvc_readonly != nil && isvc_readonly == false {
```
Comment (terrytangyuan) [inline pkg/webhook/admission/pod/storage_initializer_injector.go:259]: The issue is that you are missing a closing `}`

File: pkg/webhook/admission/pod/storage_initializer_injector.go [hunk 2]
```diff
@@ -243,6 +243,23 @@ func (mi *StorageInitializerInjector) InjectStorageInitializer(pod *v1.Pod) erro
 		// check if using direct volume mount to mount the pvc
 		// if yes, mount the pvc to model local mount path and return
 		if mi.config.EnableDirectPvcVolumeMount {
+			readonly := true
+			isvcReadonlyString, ok := pod.Annotations[constants.StorageReadonly]
```
Comment (spolti) [inline pkg/webhook/admission/pod/storage_initializer_injector.go:247]: isn't this piece duplicated?

File: pkg/controller/v1beta1/inferenceservice/controller_test.go [hunk 3]
```diff
@@ -1573,6 +1573,163 @@ var _ = Describe("v1beta1 inference service controller", func() {
...
+			// Knative service
+			actualService := &knservingv1.Service{}
+			predictorServiceKey := types.NamespacedName{Name: constants.PredictorServiceName(serviceKey.Name),
+				Namespace: serviceKey.Namespace}
+			Eventually(func() error { return k8sClient.Get(context.TODO(), predictorServiceKey, actualService) }, timeout).
+				Should(Succeed())
+
+			// TODO, make sure to check if the annotations contains the readonly annotation
+			// should be on actualService.Spec.Template.Annotations[constants.StorageReadonlyAnnotationKey]
+			Expect(actualService.Spec.Template.Annotations).
```
Comment (spolti) [inline pkg/controller/v1beta1/inferenceservice/controller_test.go:1695]: guess I forget to update this:

try
```go
Expect(actualService.Spec.Template.Annotations[constants.StorageReadonlyAnnotationKey]).
				To(Equal("false"))
```

**Comment by spolti** [comment]
@terrytangyuan this changes looks ok, do you want to take another look?
And, @harshad16 before we propose it upstream, do you mind building the kserve image and test with ODH?

**Comment by harshad16** [comment]
>  before we propose it upstream, do you mind building the kserve image and test with ODH?

@spolti do you mean to tag me here? not sure if this relate to the work i do.

**Comment by terrytangyuan** [comment]
This PR is still in draft. Is it ready for review?

**Comment by terrytangyuan** [comment]
I mentioned previously that we'll need the upstream to review this as well. Do we have a PR in upstream?

**Comment by spolti** [comment]
> > before we propose it upstream, do you mind building the kserve image and test with ODH?
> 
> @spolti do you mean to tag me here? not sure if this relate to the work i do.

Sry, wrong tag =/

**Comment by hdefazio** [comment]
> This PR is still in draft. Is it ready for review?

@terrytangyuan This is ready for review now

**Comment by hdefazio** [comment]
Closing in favor of https://github.com/kserve/kserve/pull/3885

**Pull request closed without merging**